clustermesh-apiserver/etcdinit/root.go

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package etcdinit
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"syscall"
+	"time"
+
+	"github.com/cilium/cilium/clustermesh-apiserver/etcdinit/defaults"
+	einit "github.com/cilium/cilium/pkg/etcd/init"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/spf13/cobra"
+	client "go.etcd.io/etcd/client/v3"
+)
+
+var (
+	log                     = logging.DefaultLogger.WithField(logfields.LogSubsys, "etcdinit")
+	etcdBinaryLocation      = "/usr/bin/etcd"
+	etcdDataDir             = ""
+	etcdClusterName         = ""
+	etcdInitialClusterToken = ""
+	debug                   = false
+)
+
+func NewCmd() *cobra.Command {
+	rootCmd := &cobra.Command{
+		Use:   "etcdinit",
+		Short: "Initialise an etcd data directory for use by the etcd sidecar of clustermesh-apiserver",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return InitEtcdLocal()
+		},
+	}
+	rootCmd.Flags().StringVar(&etcdDataDir, "data-dir", defaults.EtcdDataDir, "etcd data directory")
+	rootCmd.Flags().StringVar(&etcdClusterName, "cluster-name", defaults.EtcdClusterName, "etcd cluster name")
+	rootCmd.Flags().StringVar(&etcdInitialClusterToken, "initial-cluster-token", defaults.EtcdInitialClusterToken, "etcd initial cluster token")
+	rootCmd.Flags().BoolVar(&debug, "debug", false, "Debug log output")
+	return rootCmd
+}
+
+func InitEtcdLocal() error {
+	ctx, cancelFn := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancelFn()
+
+	if debug {
+		logging.SetLogLevelToDebug()
+		log.Debug("Debug logging enabled")
+	}
+
+	log := log.
+		WithField("etcdBinaryLocation", etcdBinaryLocation).
+		WithField("etcdDataDir", etcdDataDir).
+		WithField("etcdClusterName", etcdClusterName).
+		WithField("etcdInitialClusterToken", etcdInitialClusterToken)
+
+	// When the clustermesh-apiserver is launched, we create a new etcd. We don't support persistence of any kind
+	// here. So we just delete the contents of the data directory. It should be empty, but if the initialization failed
+	// we might re-run this as an init container.
+	log.Info("Deleting contents of data directory")
+	dir, err := os.ReadDir(etcdDataDir)
```

Comment:
Honestly, I just missed that function. I guess one issue is that this will delete the directory itself, whereas the looping implementation just deletes everything **inside** the directory. Considering that the directory is a mount (of an emptydir volume), will that cause errors? 🤔

Reply:
Good point. Yeah, I guess your approach is better in this case. I'd suggest to mention that in the comment, just for additional clarity.